pkg/nmi/server/redirector_linux.go

```diff
@@ -0,0 +1,83 @@
+// +build linux
+
+package server
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	iptables "github.com/Azure/aad-pod-identity/pkg/nmi/iptables"
+	"k8s.io/klog"
+)
+
+// LinuxRedirector returns sync function for linux redirector
+func LinuxRedirector(server *Server) func(*Server, chan bool, chan bool) {
+	return func(server *Server, subRoutinedone chan bool, mainRoutineDone chan bool) {
+		updateIPTableRules(server, subRoutinedone, mainRoutineDone)
+	}
+}
+
+// WindowsRedirector returns sync function for windows redirector
+func WindowsRedirector(server *Server) func(*Server, chan bool, chan bool) {
+	panic("Windows Redirector is not applicable")
+}
+
+func updateIPTableRulesInternal(server *Server) {
+	klog.V(5).Infof("node(%s) hostip(%s) metadataaddress(%s:%s) nmiport(%s)", server.NodeName, server.HostIP, server.MetadataIP, server.MetadataPort, server.NMIPort)
+
+	if err := iptables.AddCustomChain(server.MetadataIP, server.MetadataPort, server.HostIP, server.NMIPort); err != nil {
+		klog.Fatalf("%s", err)
+	}
+	if err := iptables.LogCustomChain(); err != nil {
+		klog.Fatalf("%s", err)
+	}
+}
+
+// updateIPTableRules ensures the correct iptable rules are set
+// such that metadata requests are received by nmi assigned port
+// NOT originating from HostIP destined to metadata endpoint are
+// routed to NMI endpoint
+func updateIPTableRules(server *Server, subRoutinedone chan bool, mainRoutineDone chan bool) {
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, syscall.SIGTERM, syscall.SIGINT)
+
+	ticker := time.NewTicker(time.Second * time.Duration(server.IPTableUpdateTimeIntervalInSeconds))
+	defer ticker.Stop()
+
+	// Run once before the waiting on ticker for the rules to take effect
+	// immediately.
+	updateIPTableRulesInternal(server)
+	server.Initialized = true
+
+	for {
+		select {
+		case <-mainRoutineDone:
+		case <-signalChan:
+			handleTermination()
+			subRoutinedone <- true
```

Comment:
Correct semantics for stop channels is to close them. Not send a single message to them.
Closed channel will not block in multiple calls, but with single message the channel will block for the next message.
So if reader has multiple reads from the channel it will not handle termination correctly

Reply:
Good point.